list_test.go

```diff
@@ -430,6 +430,70 @@ func TestListGetContainers(t *testing.T) {
 	assert.Equal(state, []fullContainerState(nil))
 }
 
+func TestListGetContainersMissingRootfs(t *testing.T) {
+	assert := assert.New(t)
+
+	tmpdir, err := ioutil.TempDir(testDir, "")
+	assert.NoError(err)
+	defer os.RemoveAll(tmpdir)
+
+	pod := &vcMock.Pod{
+		MockID: testPodID,
+	}
+
+	rootfs := filepath.Join(tmpdir, "rootfs")
+
+	container := vc.ContainerStatus{
+		ID: pod.ID(),
+		State: vc.State{
+			State: vc.StateRunning,
+		},
+		PID: 1,
+
+		// rootfs specified, but it doesn't exist
+		RootFs: rootfs,
+	}
+
+	testingImpl.ListPodFunc = func() ([]vc.PodStatus, error) {
+		return []vc.PodStatus{
+			{
+				ID: pod.ID(),
+				ContainersStatus: []vc.ContainerStatus{
+					container,
+				},
+			},
+		}, nil
+	}
+
+	defer func() {
+		testingImpl.ListPodFunc = nil
```

Comment:
dump question why this is need?

Reply:
This resets the virtcontainers mock handler for `ListPod()`. See https://github.com/containers/virtcontainers/blob/master/pkg/vcMock/mock.go#L91..L99.

By resetting the handler to `nil`, we guarantee that any other test that invokes any virtcontainers API calls which either directly or indirectly call `ListPod()` will fail with an error from the virtcontainers mock package. The reason for all this being that some virtcontainers types are private, but the runtime's unit tests need to create particular test scenarios using those private data types. They also need to simulate previous calls to particular sequences of virtcontainers calls. Since the runtime uses the virtcontainers API heavily, we have a problem as we cannot contrive such scenarios using the main virtcontainers API calls (the runtime tests might not even have sufficient privs to run some of the virtcontainers API calls). As such, virtcontainers provides *two* implementations of its API - the "proper" / official one and a mock one that can be used for testing the caller of the API itself, as here.